### PR TITLE
Add Ballista examples

### LINKED ISSUE #774: Add Ballista examples
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
It is hard for new users to get started.

**Describe the solution you'd like**
Examples.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes #774 .

 # Rationale for this change

Examples help new users get started.

# What changes are included in this PR?

- There are now Ballista versions of the DataFusion DataFrame and SQL examples
- The Ballista README provides an extremely brief overview of how Ballista works. This will need much more work but it is a start.

 
# Are there any user-facing changes?

No

### CODE REVIEW

**Review by houqp — APPROVED** [review]
the scheduler overview section is very clear :+1:

File: ballista-examples/examples/ballista-dataframe.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    // define the query using the DataFrame trait
+    let df = ctx
+        .read_parquet(filename)?
+        .select_columns(&["id", "bool_col", "timestamp_col"])?
+        .filter(col("id").gt(lit(1)))?;
+
+    // execute the query - note that calling collect on the DataFrame
+    // trait will execute the query with DataFusion so we have to call
+    // collect on the BallistaContext instead and pass it the DataFusion
```
Comment (alamb) [inline ballista-examples/examples/ballista-dataframe.rs:43]: 👍

File: ballista-examples/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+version = "0.5.0-SNAPSHOT"
+homepage = "https://github.com/apache/arrow-datafusion"
+repository = "https://github.com/apache/arrow-datafusion"
+authors = ["Apache Arrow <dev@arrow.apache.org>"]
+license = "Apache-2.0"
+keywords = [ "arrow", "distributed", "query", "sql" ]
+edition = "2018"
+publish = false
+
+
```
Comment (alamb) [inline ballista-examples/Cargo.toml:30]: I wonder if it is worth adding `bin` targets here?

As it is I can't run these examples:

```shell
(arrow_dev) alamb@MacBook-Pro:~/Software/arrow-datafusion/ballista-examples$ cargo run 
error: a bin target must be available for `cargo run`
```

Maybe something like
```toml
[[bin]]
name = "dataframe"
path = "src/ballista_dataframe.rs"
```

File: ballista/README.md [hunk 3]
```diff
@@ -35,9 +35,30 @@ Ballista can be deployed as a standalone cluster and also supports [Kubernetes](
...
+
+Fully working examples are available. Refer to the [Ballista Examples README](../ballista-examples/README.md) for 
+more information.
+
+## Distributed Scheduler Overview
+
+Ballista uses the DataFusion query execution framework to create a physical plan and then transforms it into a 
+distributed physical plan by breaking the query down into stages whenever the partitioning scheme changes.
+
+Specifically, any `RepartitionExec` operatoris is replaced with an `UnresolvedShuffleExec` and the child operator
```
Comment (alamb) [inline ballista/README.md:48]:
```suggestion
Specifically, any `RepartitionExec` operator is is replaced with an `UnresolvedShuffleExec` and the child operator 
```

**Review by alamb — APPROVED** [review]
LGTM -- documentation for the win!

File: ballista/README.md (hunk 3 above)
Comment (houqp) [reply ballista/README.md:48]: we have an extra is here right?

Comment (alamb) [reply ballista/README.md:48]: oh man -- good eyes. You are right -- the correct suggestion is:

```suggestion
Specifically, any `RepartitionExec` operator is replaced with an `UnresolvedShuffleExec` and the child operator 
```

File: ballista-examples/Cargo.toml (hunk 2 above)
Comment (andygrove) [reply ballista-examples/Cargo.toml:30]: I was following the same pattern that we use in `datafusion-examples` where we use `cargo run --example` rather thaan `cargo run --bin`.

```
% cargo run --example
error: "--example" takes one argument.
Available examples:
    ballista-dataframe
    ballista-sql
```

It is a little odd that we package the examples in their own crate, so maybe packaging them as binaries makes more sense now?

Comment (Dandandan) [reply ballista-examples/Cargo.toml:30]: The main motivation from extracting them into a separate folder/crate for datafusion-examples was to reduce the nr of dependencies and compilation time.
Maybe bin works just as well?

Comment (andygrove) [reply ballista-examples/Cargo.toml]: I have updated this to use `--bin` now.

Comment (alamb) [reply ballista-examples/Cargo.toml]: 🎉

**Pull request merged**